### PR TITLE
Enhance LDAP errors handling

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #14556

Meant to replace #14356

### CODE REVIEW

**Comment by trasher** [comment]
As draft for now, for first feedbacks.

File: src/Auth.php [hunk 1]
```diff
@@ -277,23 +277,39 @@ public function connection_ldap($ldap_method, $login, $password)
...
+                            sprintf(
+                                'Unable to bind %s',
+                                $dn
+                            )
+                        )
+                    );
+                }
+            }
+
+            if ($this->user_found) {
```
Comment (cedric-anne) [inline src/Auth.php:296]: IMHO, `$this->user_found` should remain to `true` even if `ldap_bind` fails. Indeed, user is still present in LDAP, as we found it, but he probably used a wrong password. Asserting that it is not anymore present in LDAP directory may lead to its deactivation/removal/....

```suggestion
            if ($this->user_found) {
                $bind_result = @ldap_bind($this->ldap_connection, $dn, $password) === false

                if ($bind_result === false) {
                    $this->addToError(
                        AuthLDAP::buildError(
                            $this->ldap_connection,
                            sprintf(
                                'Unable to bind %s',
                                $dn
                            )
                        )
                    );
                }
            }

            if ($this->user_found && $bind_result !== false) {
```

File: src/AuthLDAP.php [hunk 2]
```diff
@@ -3010,22 +3026,48 @@ public static function connectToServer(
 
         $ds = @ldap_connect($host, intval($port));
         if ($ds) {
```
Comment (cedric-anne) [inline src/AuthLDAP.php:3028]: I would probably be easier to read if error/return was triggered at this point on failure. If would prevent to have to encapsulate the whole logic in a `if ($ds) { ... }`

```php
$ds = @ldap_connect($host, intval($port));
if ($ds === false) {
    trigger_error(
        sprintf(
            "Unable to connect to LDAP server %s:%s",
            $host,
            $port
        )
    );
    return false;
}
...
```

File: src/AuthLDAP.php [hunk 3]
```diff
@@ -3010,22 +3026,48 @@ public static function connectToServer(
...
+                        static::buildError(
+                            $ds,
+                            sprintf(
+                                "Unable to start TLS connection to LDAP server %s:%s",
+                                $host,
+                                $port
+                            )
+                        ),
+                        E_USER_WARNING
+                    );
```
Comment (cedric-anne) [inline src/AuthLDAP.php:3070]: We should probably stop here.
```suggestion
                    );
                    return false;
```

**Review by cedric-anne — COMMENTED** [review]
:+1: 

This is really a great move here. We should always use defensive code when dealing with a remote service, especially when it involves dynamic data entered by the end user.

I think it could be a good idea to add tests that validates that all these failing cases are producing helpful messages, at least with `openldap`.

File: src/AuthLDAP.php [hunk 4]
```diff
@@ -1907,7 +1907,10 @@ public static function searchForUsers(
                 $sr = ldap_search($ds, $values['basedn'], $filter, $attrs, 0, -1, -1, LDAP_DEREF_NEVER, $controls);
                 if ($sr === false) {
                     trigger_error(
-                        sprintf('LDAP search failed with error (%s) %s', ldap_errno($ds), ldap_error($ds)),
+                        static::buildError(
```
[the remark this answers is not in the document]
Comment (trasher) [reply src/AuthLDAP.php:1910]: Yes, I'd like to add tests on several parts, I'll see what I can do ;)

File: tests/LDAP/AuthLdap.php [hunk 5]
```diff
@@ -1494,10 +1494,18 @@ public function testRemovedUser()
...
-        $this->boolean($synchro)->isFalse();
+        $this->when(
+            function () use ($ldap, $user) {
+                $synchro = $ldap->forceOneUserSynchronization($user);
+                $this->boolean($synchro)->isFalse();
+            }
+        )
+            ->error()
+                ->withType(E_USER_WARNING)
+                ->withMessage("Unable to bind with login cn=Manager,dc=glpi,dc=org\nerror: Can't contact LDAP server (-1)")
```
Comment (trasher) [inline tests/LDAP/AuthLdap.php:1505]:
```suggestion
                ->withMessage("Unable to bind with login `cn=Manager,dc=glpi,dc=org`\nerror: Can't contact LDAP server (-1)")
```

File: src/AuthLDAP.php [hunk 6]
```diff
@@ -3009,40 +3025,89 @@ public static function connectToServer(
...
             }
+        } else {
+            $b = true;
+        }
+
+        if ($b) {
+            return $ds;
         }
+
         return false;
```
Comment (cedric-anne) [inline src/AuthLDAP.php:3111]:
```suggestion
        return $ds;
```

File: src/AuthLDAP.php [hunk 7]
```diff
@@ -3009,40 +3025,89 @@ public static function connectToServer(
...
-                @ldap_set_option(null, LDAP_OPT_X_TLS_CERTFILE, $tls_certfile);
-            }
+        if ($ds === false) {
+            trigger_error(
+                sprintf(
+                    "Unable to connect to LDAP server %s:%s",
+                    $host,
+                    $port
+                )
+            );
```
Comment (cedric-anne) [inline src/AuthLDAP.php:3036]:
```suggestion
            );
            return false;
```

**Review by cedric-anne — APPROVED** [review]
OK for me now.

We could probably add tests, and maybe improve some messages (tests would help to see what we will push in logs).

**Pull request merged**